internal/pkg/v2/infrastructure/redis/device.go

```diff
@@ -0,0 +1,82 @@
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package redis
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/edgexfoundry/edgex-go/internal/pkg/common"
+	"github.com/edgexfoundry/go-mod-core-contracts/errors"
+	model "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+
+	"github.com/gomodule/redigo/redis"
+)
+
+const DeviceCollection = "v2:device"
+
+// deviceStoredKey return the device's stored key which combines the collection name and object id
+func deviceStoredKey(id string) string {
+	return fmt.Sprintf("%s:%s", DeviceCollection, id)
+}
+
+// deviceExistByName whether the device exists by name
+func deviceExistByName(conn redis.Conn, name string) (bool, errors.EdgeX) {
+	exists, err := redis.Bool(conn.Do(HEXISTS, DeviceCollection+":name", name))
+	if err != nil {
+		return false, errors.NewCommonEdgeX(errors.KindDatabaseError, "device existence check by name failed", err)
+	}
+	return exists, nil
+}
+
+// deviceExistById checks whether the device exists by id
+func deviceExistById(conn redis.Conn, id string) (bool, errors.EdgeX) {
+	exists, err := redis.Bool(conn.Do(EXISTS, deviceStoredKey(id)))
+	if err != nil {
+		return false, errors.NewCommonEdgeX(errors.KindDatabaseError, "device existence check by id failed", err)
+	}
+	return exists, nil
+}
+
+// addDevice adds a new device into DB
+func addDevice(conn redis.Conn, d model.Device) (model.Device, errors.EdgeX) {
+	exists, edgeXerr := deviceExistById(conn, d.Id)
+	if edgeXerr != nil {
+		return d, errors.NewCommonEdgeXWrapper(edgeXerr)
+	} else if exists {
+		return d, errors.NewCommonEdgeX(errors.KindDuplicateName, fmt.Sprintf("device id %s exists", d.Id), edgeXerr)
+	}
+
+	exists, edgeXerr = deviceExistByName(conn, d.Name)
+	if edgeXerr != nil {
+		return d, errors.NewCommonEdgeXWrapper(edgeXerr)
+	} else if exists {
+		return d, errors.NewCommonEdgeX(errors.KindDuplicateName, fmt.Sprintf("device name %s exists", d.Name), edgeXerr)
```

Comment:
change `%s exists` to `%s already exists`

Reply:
Modified.